### ISSUE TITLE
Add argument to ignore deleted service account keys

### ISSUE THREAD

**Issue by ericnorris** [hashicorp/terraform-provider-google#8300]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

I have a `google_service_account_key` resource that I create as a part of bootstrapping a Terraform Cloud workspace and Google Cloud Platform project. An automated system later rotates that service account key on some cadence, which involves deleting the service account key and replacing the contents of the Terraform Cloud workspace variable.

Fortunately, once this happens the next time `terraform plan` is run in the workspace that contains the `google_service_account_resource`, [this piece of code](https://github.com/hashicorp/terraform-provider-google/blob/4373d10c86d06498cb2297bbe4c1ddf0a1309140/google/resource_google_service_account_key.go#L160-L167) catches that the key is missing and doesn't raise an error.

Unfortunately for my use case, this means that it tries to create the service account key again, which, while I can workaround in the rest of the Terraform configuration (with an `ignore_changes` block), is not ideal because I'd rather not have valid service account key material stay in my Terraform state. It also causes unneeded visual noise in the plan.

**tl;dr:** would it be possible to add an argument to the `google_service_account_resource` to have it ignore a missing key? I would expect the resource to behave similarly to how it does with the `private_key` field, which is no longer returned in the `Projects.ServiceAccounts.Keys.Get` call - it could continue returning the historical values of the key, but with the understanding that when deleted, the private key will no longer be valid.

### New or Affected Resource(s)

* google_service_account_key

### Potential Terraform Configuration

```tf
resource "google_service_account_key" "key" {
  service_account_id = google_service_account.account.id

  # not tied to this argument name
  ignore_deleted = true
}
```

### Note

If the maintainers are okay with this approach, I would be more than happy to attempt a PR.

**Comment by slevenick** [comment]
It seems like you don't want the service account key managed declaratively. Could you remove the service account key from your Terraform config and remove it from state? I'm wary about this, as it would create a very strange partially deleted state

**Comment by ericnorris** [comment]
@slevenick Thanks for the quick response!

I'd like the key _created_ declaratively, but afterwards I do not care about it. Certainly I could remove it from the config (or reduce the `count` to zero), but for my use case this is non-ideal.

I am setting up something similar to https://github.com/terraform-google-modules/terraform-google-project-factory/, but specific to my organization, and Terraform Cloud. The expected user would be someone that is trying to create a new GCP project, by adding another instance of this module to a centralized location.

In order to ensure that the newly set up project has credentials from the very beginning, I'd like to create a `google_service_account_key`. Since an external program will later rotate the key, I would no longer care about the key once it has been deleted. It would be an additional burden for someone to have to go back and manually edit the state or mess with the module configuration to tell it to no longer create the service account key, when in reality I simply want it to be a "create-only" resource.

I understand the concern, but I don't think it would create a strange state - the private key would be invalid, yes, but it's the same whether the key is in the Terraform state or if you had copy-pasted the value into a file or some service. Changing `ignore_deleted` back to `false` reverts back to the original behavior, with no lasting side effects.

**Comment by ericnorris** [comment]
I should note: I had originally tried this with a `lifecycle_rule`, but unfortunately that does not cover this. I think this is spiritually the same as if `lifecycle_rule` worked this way - you want a resource to be created by Terraform, but something else will manage it.

**Comment by slevenick** [comment]
I don't believe we want to support this option within the provider. I would suggest using an alternative method of creating the service account key if you don't want to manage the lifecycle of the resource, as Terraform is designed to manage the entire lifecycle.

You could try creating the key through gcloud or a similar imperative tool and using a variable to reference it within the Terraform config.

The problem with this solution from the provider side is that it fundamentally changes the way Terraform interacts with this GCP resource by altering the read functionality. It may make sense for Terraform core to implement a solution for this in the future to allow for create-only resources, but at this time I don't believe we should support this in the provider.

**Comment by ericnorris** [comment]
@slevenick I understand, but that answer is dissatisfactory for a number of reasons.

Requiring the use of `gcloud` prevents making this a secure (i.e. not using `local-exec`) self contained module, which is important for less-savvy end-users. Alternatively this means we will need to grant access to _humans_ in order to run `gcloud`, and involve them in a tedious process of generating initial keys every time someone wants to make a new project.

I also don't think this is unheard of for a Terraform resource, see https://github.com/hashicorp/terraform-provider-google/blob/master/google/resource_google_project_default_service_accounts.go#L22 which does absolutely nothing after the resource is initially created. Or https://github.com/hashicorp/terraform-provider-google/blob/master/google/resource_google_project.go#L180-L193, which only happens at the time of project creation.

The `google_service_account_key` resource has also added features like `keepers` in order to facilitate use-cases that are outside of just creating a service account key, without requiring a core Terraform change (arguably Terraform core could and should support the idea of `keepers`). This request feels similar, again, in trying to customize exactly when a service account key is created.

I'd appreciate it if you could reconsider, but feel free to close the issue if this is not convincing. Thanks!

**Comment by ericnorris** [comment]
As an alternative, would you be amenable to a `google_throwaway_service_account_key` resource, which implements `Create` and `Delete`, and leaves all other operations as `schema.Noop`? Not tied to that exact name, of course.

This could potentially be more explicit that it is meant for the situations I'm describing - bootstrapping something in Terraform with a service account key that is expected to be rotated (managed) outside of Terraform. Perhaps that alleviates some of your concerns with the state of the key after creation?

**Comment by slevenick** [comment]
We talked this over and while a separate resource would be a better solution than adding this directly to service account key, we don't believe this functionality is a good addition to Terraform.

There do exist examples of this type of functionality in the provider already, but we don't want to add new instances of it

Overall if a resource is declared in a Terraform config we should try to manage its lifecycle and recreate it if it has been destroyed out of band.
